### PR TITLE
feat(tree): add ancestor selection mode, input-enabled prop for selectable trees

### PR DESCRIPTION
**Pull request by paulcpederson**
## Summary

We have several UI's which are hierarchical structures where every level can be selected. Here are some examples:

Screen Shot 2021-03-16 at 6 18 31 PM

Screen Shot 2021-03-16 at 6 20 23 PM

I've added this functionality as a mode to the Tree component. `selection-mode="ancestors"` will select all children  of an object when select, and deselect all children when deselected. This selection mode is very similar to the `ancestors` selection mode that is now available in `combobox`.

I've also added a `checkboxes` prop. This works much like `lines`. It's a more form-like way to render tree options for selecting, rather than just navigating. When you have checkboxes and you're in ancestors selection mode, the tree will manage setting indeterminate states for you as well.

I've added a tests for the selection mode. I didn't convert `tree` to tailwind or write tests for the other tree functionality (tests are currently very limited on tree). Can try to take on more of the tree stuff in a follow up.

---

cc @bstifle for design 👀

### CODE REVIEW

**Comment by paulcpederson** [comment]
also fyi @capeoples

**Review by eriklharper — APPROVED** [review]

**Review by macandcheese — APPROVED** [review]
Looks good! There's a small "focus ring" issue with the checkbox on mousedown, but that very well could be an issue with the checkbox component or something existing in tree - I have generally found there to be some weird interactions between tree-items, anchor links, and navigation with mouse or keyboard.


Screen Shot 2021-03-17 at 10 36 00 AM


- Should "checkbox" display a radio when in selection-mode="single"?
- Some general spacing stuff that will hopefully be resolved with a Tailwind class. 

Thanks for adding @paulcpederson

**Comment by paulcpederson** [comment]
> Should "checkbox" display a radio when in selection-mode="single"

This is an interesting thought. We could use the `tile-select-group` of `input-enabled` plus `type`?

File: src/components/calcite-tree/calcite-tree.tsx [hunk 1]
```diff
@@ -37,6 +37,9 @@ export class CalciteTree {
   /** Display indentation guide lines */
   @Prop({ mutable: true, reflect: true }) lines = false;
 
+  /** Display checkboxes */
+  @Prop({ mutable: true }) checkboxes = false;
```
[the remark this answers is not in the document]
Comment (paulcpederson) [reply src/components/calcite-tree/calcite-tree.tsx:41]: I think I might change this to `input-enabled` which is an adjective and consistent with `tile-select-group`

File: src/components/calcite-tree-item/calcite-tree-item.scss [hunk 2]
```diff
@@ -110,6 +114,17 @@
   @include focus-style-inset();
 }
 
+.calcite-tree-checkbox {
+  line-height: 0;
```
[the remark this answers is not in the document]
Comment (paulcpederson) [reply src/components/calcite-tree-item/calcite-tree-item.scss:118]: Yeah I figured the tailwind pass would have to come in a later pr. There is a ton of standard CSS at the moment.

**Comment by paulcpederson** [comment]
Addressed most of the feedback.

Updated to input-enabled. We can make it use radios for the single select modes in the future.

**Comment by bstifle** [comment]
is there a reason we are adding a new style with checkboxes? i thought we always had a design with checkmarks

image

**Comment by paulcpederson** [comment]
@capeoples can probably speak to this requirement more, but having a mixed state with icons is harder. And I think while heavier, the checkbox version is more form-like.

**Comment by capeoples** [comment]
@bstifle, we have cases in Online where both options would be useful. 

I think the style as it is right now is great for filters or selectors in a combobox. 

As @paulcpederson said, having a checkbox affordance is nice for forms, or in cases where you a selecting content.

**Comment by bstifle** [comment]
cool beans. we'll get that in the figma

**Comment by bstifle** [comment]
so are we calling this `selection-mode="ancestors"` ?

**Comment by paulcpederson** [comment]
@bstifle yeah it's two separate props, the `selection-mode="ancestors"` controls the selection up/down the tree, while `input-enabled` shows the checkbox

**Pull request merged**